### PR TITLE
Add "Open Source Maintenance" blog post

### LINKED ISSUE #244: Blog Post: Open Source Maintenance
this post should describe how to speed up work on open source and other projects
- shell aliases for easy git cloning, pushing and opening PRs
- having tests
- running tests in CI (e.g. TravisCI)
- using a dependency update bot (e.g. @dependabot)
- using a changelog generator (e.g. https://github.com/lerna/lerna-changelog)

### PR DESCRIPTION
**Pull request by Turbo87**
Resolves #244 

Rendered preview is available at https://github.com/Turbo87/simplabs.github.io/blob/osm/blog/_posts/2018-11-27-open-source-maintenance.md (tables don't seem to render properly there though...)

### CODE REVIEW

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+Node.js 6, then you should increase the **major** version.
+
+
+#### Dependency Update Services
+
+Any sufficiently large open source project has at least a few dependencies that
+it is built upon, and even smaller projects typically have at least a dependency
+on a test framework. While keeping the dependencies of a single project
+up-to-date is a manageable task, it is becoming a full-time job once you
+maintain dozens or hundreds of separate projects.
```
Comment (geekygrappler) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: Dozens or hundreds sounds ambitious to me, which might detract from the point, like I should only consider dependency update service once I pass the 50 OSS project mark :trollface: . 
Maybe:
> While keeping the dependencies of a single  project up-to-date is  a manageable task, it quickly becomes a full-time job once you maintain multiple separate projects.

**Review by geekygrappler — APPROVED** [review]

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 1 above)
Comment (Turbo87) [reply blog/_posts/2018-11-27-open-source-maintenance.md]: https://github.com/Turbo87?tab=repositories ... tell me about it 😆 

but yeah, I see what you mean. will update that phrasing.

**Review by oliverbarnes — APPROVED** [review]
👏 another great post, and I learned about Fork, which I'm trying out now :)

only comment I have, might sound familiar - would be good to mention ruby and elixir tooling together with examples for rust or python, given they're languages we offer services for and have open source projects written in. it may sound markety, but I think it does make a difference with clients. Not to mention it's good SEO =D

**Comment by Turbo87** [comment]
@oliverbarnes I totally agree, unfortunately I personally don't have any experience with these, so I didn't know what tools to recommend. do you have any specific tools in mind that I should mention?

**Comment by oliverbarnes** [comment]
Sure:

**ruby**

- testing: [rspec](http://rspec.info/) and [test-unit](https://github.com/test-unit/test-unit) (ruby has a strong testing culture, rails has it built-in)

- package management: [bundler](https://bundler.io/) (follows semver but doesn't have testing built-in like `cargo`)

- linting: [rubocop](https://github.com/rubocop-hq/rubocop)


**elixir**

- testing: [ExUnit](https://hexdocs.pm/ex_unit/ExUnit.html)

- package management: [Mix](https://hexdocs.pm/mix/Mix.html) (semver and testing built-in)

- linting: not so familiar with what's out there these days, but I found [dogma](https://github.com/lpil/dogma) (motto: _A code style linter for Elixir, powered by shame_) and [credo](https://github.com/rrrene/credo) to be the github repos with most popularity

...ah, and unrelated to the ruby/elixir tooling thing: 

about CI options, might be worth mentioning [Gitlab's built-in CI](https://about.gitlab.com/product/continuous-integration/)

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+<!--break-->
+
+<style>
+code {
+  white-space: nowrap;
+}
+</style>
+
+#### Git
```
Comment (ghislaineguerin) [inline blog/_posts/2018-11-27-open-source-maintenance.md:23]:
```suggestion
## An optimized workflow for maintainers
```
Adding this for better keyword placement.

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+Version control systems have always been an important part of professional
+software development, but the development of [git] over a decade ago and a
+short while later [GitHub] has made version control mainstream for open source
+projects.
+
+While a lot of tutorials focus on using git from the command line, we have found
```
Comment (ghislaineguerin) [inline blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
While many tutorials focus on using git from the command line, we have found
```

**Review by ghislaineguerin — COMMENTED** [review]
@Turbo87 Looks good. Few tips to make this more valuable for SEO:

### Mention common roles involved in open-source like Contributors, Maintainers
These roles are closely related to open source projects topic and will help with ranking.

### Apps. vs Open-source projects
We only mention open source projects on the first two paragraphs, replacing our references to apps and including open source will help improve our topic keyword density.

### Best Practices 
This is a very common search term for people looking to improve their workflows. Try to mention it somewhere in the article, better if it's a full sentence like Best practices for open source projects maintenance.

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+that using a graphical user interface to visualize the history graph can make
+it quite a bit easier for beginners to understand what is going on. One good
+example of such a tool is [Fork], which is freely available for Windows and Mac:
+
+![Screenshot of Fork](/images/posts/2018-11-27-open-source-maintenance/git-fork.png)
+
+When working on multiple different projects it is very useful to rely on similar
```
Comment (ghislaineguerin) [inline blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
When working on multiple different projects, it is useful to rely on similar
```

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+conventions, for example when naming things. In the context of git this is most
```
Comment (ghislaineguerin) [inline blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
conventions, for example, when naming things. In the context of Git this is most
```

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 3 above)
Comment (ghislaineguerin) [inline blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
While a lot of tutorials focus on using Git from the command line, we have found
```

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+Most of the time we use [TravisCI] for this purpose, which is coupled to GitHub,
+free for open source projects, and supports Linux, Mac and soon Windows too.
+Alternatives include [CircleCI] (based on Docker images) and [AppVeyor], which
+we currently use to test Windows support on projects where this is relevant.
+
+One important thing to mention here is that CI can not only run your test suite
+but can often also do other things like publishing new versions of your
+projects. On most of our projects we have configured TravisCI to automatically
+publish new releases to [npm] whenever we push a new git tag to the server.
```
Comment (ghislaineguerin) [inline blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
publish new releases to [npm] whenever we push a new Git tag to the server.
```

**Review by jorgelainfiesta — APPROVED** [review]
Nice article, it was easy to follow even for me who have no experience maintaining an open source project.

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 3 above)
Comment (Turbo87) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: merging the suggestions because GitHub can't do it on its own... 😅 

```suggestion
While many tutorials focus on using Git from the command line, we have found
```

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+Version control systems have always been an important part of professional
+software development, but the development of [git] over a decade ago and a
+short while later [GitHub] has made version control mainstream for open source
+projects.
+
+While many tutorials focus on using Git from the command line, we have found
+that using a graphical user interface to visualize the history graph can make
+it quite a bit easier for beginners to understand what is going on. One good
+example of such a tool is [Fork], which is freely available for Windows and Mac:
```
Comment (marcoow) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: Strictly, it's _"macOS"_ when you want to refer to the OS.

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+Similar to testing frameworks, a lot of ecosystems have tools to run "static
+analysis" on your code to find bugs before you even run the applications. These
+tools are often called "Linters".
+
+Very popular in the JavaScript world is [ESLint], which we often run in
+combination with [Prettier], to also ensure that our code is formatted in a
+consistent way.
+
+Again, similar tools exist for example for Rust ([clippy] and [rustfmt]) and
+Python ([pyflakes] and [black]).
```
Comment (marcoow) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: Would be good to mention the `mix format [--check-formatted]` task for Elixir as well (could also be instead of Python that we don't really have a connection with).

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+   users you should increase the **major** version
+3. when your release includes only bug fixes you should increase the **patch**
+   version
+4. for all other changes you should increase the **minor** version
+5. if your version is below `v1.0.0` (e.g. `v0.4.3`) you can release breaking
+   changes without increasing the major version because the project is
+   considered "unstable"
+
+Since the majority of package managers including [npm], [yarn], [cargo] and
+[pip] follow these semantics when resolving dependencies, it is very important
```
Comment (marcoow) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: would be good to include https://hex.pm as well - again, could be instead of pip

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+to comply with these rules. But they are also useful in order for your users
+to get a sense of what they can expect from a new release.
+
+While this can be controversial, we consider dropping support for older Node.js,
+Python, Ruby, or other language releases a breaking change. This means that if
```
Comment (marcoow) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: having Elixir (and maybe Rust) in here would be good, I'd replace Python with it

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+up-to-date is a manageable task, it is becoming a full-time job once you
+maintain dozens or hundreds of separate projects.
+
+Fortunately for us this is a task that can be automated quite well, at least
+if you have a good test suite and continuous integration set up. While we
+first used [Greenkeeper] for this task, we have lately transitioned to using
+[dependabot], which supports not only npm, but all sorts of different package
+managers, language ecosystems, and even monorepos.
+
+The way this work is that dependabot will automatically open Pull Requests on
```
Comment (marcoow) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: _"…this work**s**…"_

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+While "semantic commit messages" seem to be quite popular we prefer the second
+group of changelog generators, as listing each commit often results in a long
+list of very fine-grained changes that are not directly helpful to the user.
+
+Instead we use [lerna-changelog] to generate our changelogs from all of the
+merged pull requests that went into each of the releases. To make this work
+properly it is important to focus each pull request on only one single, logical
+change, and label it properly as either `enhancement`, `bug`, `breaking`, or
+any of the other supported/configured labels. To ensure that all of our projects
+use the same set of labels we use [github-label-sync].
```
Comment (marcoow) [inline blog/_posts/2018-11-27-open-source-maintenance.md:227]: is this true? 😆

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 7 above)
Comment (marcoow) [reply blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
example of such a tool is [Fork], which is freely available for Windows and macOS:
```

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 8 above)
Comment (marcoow) [reply blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
Elixir (`mix format [--check-formatted]`).
```

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 9 above)
Comment (marcoow) [reply blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
[hex](https://hex.pm) follow these semantics when resolving dependencies, it is very important
```

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 10 above)
Comment (marcoow) [reply blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
Elixir, Ruby, Rust, or other language releases a breaking change. This means that if
```

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 11 above)
Comment (marcoow) [reply blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
The way this works is that dependabot will automatically open Pull Requests on
```

**Review by marcoow — COMMENTED** [review]
Nice post!

I think it's fine but not necessary to present everything in here as sort of the official simplabs way of doing things - e.g. the aliases etc. are very likely to be different for everyone (I don't use any at all for example).

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 12 above)
Comment (Turbo87) [reply blog/_posts/2018-11-27-open-source-maintenance.md:227]: at least for the projects that I manage... 😄

Comment (Turbo87) [reply blog/_posts/2018-11-27-open-source-maintenance.md:227]: see https://github.com/Turbo87/github-labels

Comment (marcoow) [reply blog/_posts/2018-11-27-open-source-maintenance.md:227]: 👍 would be good to set up a simplabs/github-labels as well

Comment (Turbo87) [reply blog/_posts/2018-11-27-open-source-maintenance.md:227]: feel free to fork it. my set of labels happens to match the default labels of lerna-changelog. 😉 

see https://github.com/lerna/lerna-changelog/blob/master/src/configuration.ts#L53-L57

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+code {
+  white-space: nowrap;
+}
+</style>
+
+#### Git
+
+Version control systems have always been an important part of professional
+software development, but the development of [git] over a decade ago and a
+short while later [GitHub] has made version control mainstream for open source
```
Comment (mansona) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: Maybe instead of

> but the development of [git] over a decade ago and a short while later [GitHub]

Something like

> but the development of [git] over a decade ago and the subsequent development of [GitHub]

Might sound a bit better? I'm not sure though 🤔

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+When working on multiple different projects, it is useful to rely on similar
+conventions, for example, when naming things. In the context of Git this is most
+relevant when naming [remotes](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes)
+and branches.
+
+For branches we use the default `master` branch as our main development branch.
+Feature branches are named after the thing that they are implementing or fixing
+and ideally only contain isolated, focused changes. If we publish new releases
+we "tag" the release commits with a tag name like `v1.2.3`, corresponding to
```
Comment (mansona) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: Do you use [np](http://npmjs.com/package/np)? might be worth mentioning that there are tools that can do these tags and release commits for you but it might not fit in this section of the blog 🤔

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+crate.
+
+
+#### Linting
+
+Similar to testing frameworks, a lot of ecosystems have tools to run "static
+analysis" on your code to find bugs before you even run the applications. These
+tools are often called "Linters".
+
+Very popular in the JavaScript world is [ESLint], which we often run in
```
Comment (mansona) [inline blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
A very popular linter in the JavaScript world is [ESLint], which we often run in
```

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 16]
```diff
@@ -0,0 +1,257 @@
...
+#### Continuous Integration
+
+The above points about testing and linting are nice to have, but if nobody runs
+your tests then they don't provide any value. This is where continuous
+integration (or short "CI") systems come into play. These kinds of systems are
+coupled to your version control servers and automatically run linting checks and
+tests whenever you upload new commits to the server.
+
+Most of the time we use [TravisCI] for this purpose, which is coupled to GitHub,
+free for open source projects, and supports Linux, Mac and soon Windows too.
```
Comment (mansona) [inline blog/_posts/2018-11-27-open-source-maintenance.md]: I would rephrase this from 

> Most of the time we use [TravisCI] for this purpose, which is coupled to GitHub, free for open source projects, and supports Linux, Mac and soon Windows too.

To something more like: 

> Most of the time we use [TravisCI] for this purpose which easily integrates with GitHub, is free for open source projects, and supports running your tests in Linux and Mac Environments and will soon support Windows too.

It's a bit more verbose but the previous one felt a bit stilted to me 🤔

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 17]
```diff
@@ -0,0 +1,257 @@
...
+Alternatives include [CircleCI] (based on Docker images) and [AppVeyor], which
+we currently use to test Windows support on projects where this is relevant.
+
+One important thing to mention here is that CI can not only run your test suite
+but can often also do other things like publishing new versions of your
+projects. On most of our projects we have configured TravisCI to automatically
+publish new releases to [npm] whenever we push a new Git tag to the server.
+You can find instruction on how to configure this in their official
+[documentation](https://docs.travis-ci.com/user/deployment/npm/).
```
Comment (mansona) [inline blog/_posts/2018-11-27-open-source-maintenance.md:145]: Should we be recommending auto-deployment at the moment considering the npm security issues going around? Shouldn't we be encouraging people to use 2fa for all npm deployments?

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+up-to-date is a manageable task, it is becoming a full-time job once you
+maintain dozens or hundreds of separate projects.
+
+Fortunately for us this is a task that can be automated quite well, at least
+if you have a good test suite and continuous integration set up. While we
+first used [Greenkeeper] for this task, we have lately transitioned to using
+[dependabot], which supports not only npm, but all sorts of different package
+managers, language ecosystems, and even monorepos.
+
+The way this works is that dependabot will automatically open Pull Requests on
```
Comment (mansona) [inline blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
Dependabot will automatically open Pull Requests on
```

File: blog/_posts/2018-11-27-open-source-maintenance.md [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+your projects, whenever one of your dependencies has published a new version.
```
Comment (mansona) [inline blog/_posts/2018-11-27-open-source-maintenance.md]:
```suggestion
your projects whenever one of your dependencies has published a new version.
```

**Review by mansona — APPROVED** [review]
Love the post 😍 My comments/suggestions are really minor and you can take them or leave them 👍

**Review by ghislaineguerin — APPROVED** [review]

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 14 above)
Comment (Turbo87) [reply blog/_posts/2018-11-27-open-source-maintenance.md]: I usually use `npm/yarn version` for that. yeah, might be valuable to mention that somewhere 🤔

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 17 above)
Comment (Turbo87) [reply blog/_posts/2018-11-27-open-source-maintenance.md:145]: It's a matter of trade-offs. The recent situation wouldn't have been prevented by forbidding auto-deploy or 2FA, because the keys to the kingdom were deliberately passed over. At the same time there is also a security risk involved in giving npm ownership to a large number of developers (compared to only a single account for CI). Another risk is publishing some private files in the repo (e.g. `.env` with secret keys) if `.npmignore` (or `files` in `package.json`) is not set up properly.

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 14 above)
Comment (mansona) [reply blog/_posts/2018-11-27-open-source-maintenance.md]: I'm pretty sure np uses npm/yarn version under the hood, but it also does a few other things that stop me shooting myself in the foot all the time 😂

**Comment by Turbo87** [comment]
@marcoow blog post is updated with all of the feedback and ready for another round of review and/or merging :)

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 2 above)
Comment (marcoow) [reply blog/_posts/2018-11-27-open-source-maintenance.md:23]: @Turbo87 I'd generally ask not to resolve conversations without giving the original commenter feedback

**Review by marcoow — APPROVED** [review]

File: blog/_posts/2018-11-27-open-source-maintenance.md (hunk 2 above)
Comment (Turbo87) [reply blog/_posts/2018-11-27-open-source-maintenance.md:23]: the section that follows this heading is specifically about Git and version control systems, so I prefer to keep the current heading. I've added the "Best Practices" suggestion to the above-the-fold text.
